ads/google/test/test-doubleclick.js

```diff
@@ -0,0 +1,363 @@
+      });
+    }).then(pubads => {
+      expect(pubads.get('page_url')).to.equal(config.pageUrl);
+      const slot = canvas.slot;
+      expect(slot).to.not.be.null;
+      return poll('Actual ad loaded', () => {
+        return canvas.querySelector(
+            '[id="' + config.adIframeId + '"]');
+      }, null, 5000);
+    }).then(adIframe => {
+      expect(ampAdIframe.contentWindow.context.hidden).to.be.false;
+      return poll('Creative content loaded', () => {
+        return adIframe.contentWindow.document.body.childNodes !== 0;
+      }, null, 5000);
+    }).then(() => {
+      expect(ampAdIframe.getAttribute('width')).to.equal(
+          config.primarySize.width);
+      expect(ampAdIframe.getAttribute('height')).to.equal(
+          config.primarySize.height);
+      if (config.expectResize) {
+        ampAdIframe.contentWindow.scrollTo(0,
+            ampAdIframe.contentWindow.document.body.scrollHeight);
+      }
+      return poll('Creative id transmitted. Ad fully rendered.', () => {
+        return ampAd.creativeId;
+      }, null, waitTime);
+    }).then(creativeId => {
+      expect(creativeId).to.match(/^dfp-/);
+
+      // If a secondary dimension is larger than a larger dimension, the
+      // returned creative ought to be empty.
+      if (config.expectEmpty) {
+        expect(creativeId).to.equal('dfp-_empty_');
+      }
+      if (creativeId == 'dfp-_empty_') {
+        // Nothing else to do; bail out of test.
+        return null;
+      }
+      const canvasRect = canvas.getBoundingClientRect();
+      expect(canvasRect.width).to.equal(
+          Number(config.actualCreativeSize.width));
+      expect(canvasRect.height).to.equal(
+          Number(config.actualCreativeSize.height));
+      if (!config.expectResize) {
+        return null;
+      }
+      return poll('<amp-ad> resized', () => {
+        const newAmpAdStyle = ampAd.getAttribute('style');
+        if (newAmpAdStyle != initialAmpAdStyle) {
+          return newAmpAdStyle;
+        }
+      }, null, 15000);
+    }).then(newAmpAdStyle => {
+      if (config.expectResize) {
+        expect(newAmpAdStyle).to.not.be.null;
+      }
+      if (newAmpAdStyle) {
+        expect(newAmpAdStyle).to.contain('width: ' +
+            config.actualCreativeSize.width);
+        expect(newAmpAdStyle).to.contain('height: ' +
+            config.actualCreativeSize.height);
+      }
+    });
+  });
+};
+
+const htmlFirstHalf =
+'<!doctype html>' +
+'<html ⚡>' +
+'  <head>' +
+'    <meta charset="utf-8">' +
+'    <title>DoubleClick Test</title>' +
+'    <link rel="canonical" href="http://nonblocking.io/" >' +
+'    <script async src="https://cdn.ampproject.org/v0.js"></script>' +
+'    <meta name="amp-3p-iframe-src" ' +
+'content="http://localhost:9876/dist.3p/current/frame.max.html">' +
+'  </head>' +
+'  <body>';
+
+const htmlSecondHalf = '</body></html>';
+
+const createFixtureHtml = function(config, fold) {
```

Comment:
this will be very hard to code and maintain.
could you add the HTML template to `/test/fixtures/`, and try to embed dynamic content using string replace?